chromium: Support building with NaCl

sivertb opened the pull request:
If enableNaCl is true, the NaCl and PNaCl toolchains are downloaded and
patched to use the correct dynamic linker. It also disables stripping of
the installed binaries, as this seem to cause segfaults when trying to
run NaCl apps.

sivertb commented:
Tested building with enableNaCl set to both true and false with the stable channel.

sivertb commented:
cc @aszlig

aszlig commented:
What's the reason to switch to `stdenv_32bit` for Chromium? Also, I think we really want to just exclude stripping NaCl binaries, not all binaries.

aszlig commented:
IMO it would make more sense to use mkChromiumDerivation instead of building NaCl together with the main derivation of the browser, which not only keeps things separate (and/or breaks the main browser for the sake of having NaCl support) but also doesn't cause a full rebuild of the main browser if someone wants to have NaCl support.

sivertb commented:
I switched to stdenv_32bit, because some of the toolchain binaries were 32bit, so we need $NIX_CC/nix-support/dynamic-linker-m32 to patch it.

Not sure I follow your point about mkChromiumDerivation. Do you mean build only the NaCl plugin in a separate derivation, and keep the main browser common?

sivertb commented:
Rebased, added stripping of everything except .nexe file, and changed to only add disable_glibc=1 if we're building with NaCl.

I also looked into the possibility of moving the NaCl stuff out of common.nix, but I'm not sure it's possible. If NaCl is to be used, the main browser must be built with it. I do not think it is possible to build the browser without NaCl, then drop in the NaCl runtime + plugin later.

aszlig commented:
I'm pretty sure it's possible to keep NaCl separate from the main browser, even if it's just the NaCl helpers. Going to look into that by the end of the week.

sivertb commented:
I did originally download the NaCl toolchains separately, and specified the toolchain path with `pnacl_newlib_toolchain` as described in http://www.chromium.org/nativeclient/pnacl/building-pnacl-components-for-distribution-packagers, but I ran into some trouble with stamp files. Can't recall if the trouble was with version 42 (no longer a problem) or newer. It would probably be possible to get this working with some fiddling around.

zimbatm commented:
Does it mean that chromium is forced to be 32bit because of NaCL ? I will close this issue if that's the case.

sivertb commented:
No, the produced binary is 64bit on x86_64. It's some of the downloaded toolchains that are 32bit, and require the 32bit linker.

aszlig commented:
I'm currently refactoring the Chromium build, so I'd also like to get this in. However, did you try this with enabled Nix sandboxing, because it seems that it is trying to fetch stuff during build.

sivertb commented:
Yeah, I just confirmed that it fails with build-use-chroot=true. It seems the toolchains have to be split out into a fixed-output derivation. The solution I have in mind is:

- Add
source.toolchain which simply runs the script to download and unpack the toolchains, putting them into their own derivation
- Change update.nix / update.sh to also calculate the hash of this derivation by downloading and running the scripts, then nix-store --dump .. | sha256sum

This is a bit ugly, as it will duplicate the downloading / unpacking of the toolchains into both the source.toolchain derivation and the update.sh script. If anyone has a better suggestion I'm all ears. I won't have time to look any closer at this the coming week, but I hope to have some time the week after.

sivertb commented:
Updated the change to build the toolchains as a separate derivation. It's now possible to build with build-use-sandbox=true.

aszlig commented on pkgs/applications/networking/browsers/chromium/update.sh:
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh -e
 cd "$(dirname "$0")"
-sp="$(nix-build -Q --no-out-link update.nix -A update)"
+sp="$(nix-build -Q --no-out-link update.nix -A update --option build-use-sandbox false)"
```
This won't have *any* effect if you're running the Nix daemon.

sivertb replied:
I assume update.sh is run manually, not through the nix daemon no? Anyway, I see that update.nix already fetches the tarballs without any hashes, so I assume this is already run outside a sandbox.

aszlig commented on pkgs/applications/networking/browsers/chromium/update.nix:
```diff
@@ -51,6 +51,11 @@ in rec {
       sha256 = chanAttrs.sha256bin32;
     } else throw "No Chrome plugins are available for your architecture.");
+
+    toolchain = import ./toolchain.nix {
```
So you're fetching the source to `$out` of the `toolchain.nix`, which basically implies that you need to "turn off sandboxing temporarily to fetch the toolchain" and afterwards turn on to build it along with the browser itself, do I understand this correctly?

If yes, this shouldn't be the way to go but we should rather look into what the toolchain fetcher is doing and do the packaging in a way it doesn't require sandboxing to be disabled.

aszlig replied:
Ah, that's for pre-seeding the fixed-output derivation... that's quite fragile TBH and as said, I'd really check what that toolchain fetcher is doing and replicate that behavior instead of hoping that the unsandboxed fetcher will be deterministic.

aszlig commented on pkgs/applications/networking/browsers/chromium/browser.nix:
```diff
@@ -7,14 +7,28 @@ mkChromiumDerivation (base: rec {
+  # Trying to run NaCl apps with a stripped nacl_irt_*.nexe file results in a
+  # segfault. We therefore disable stripping here and manually strip the
+  # binaries when they're copied.
+  dontStrip = true;
```
As mentioned before, we *really* should build NaCl as a separate derivation, that's why we have the `mkChromiumDerivation`. The NaCl derivation could then be referenced either via the wrapper or from the browser derivation, although I'd prefer the latter because not everybody will enjoy recompiling the main browser *and* NaCl just for the sake to have NaCl support.

aszlig commented on pkgs/applications/networking/browsers/chromium/common.nix:
```diff
@@ -149,6 +149,10 @@ let
+    '' + optionalString enableNaCl ''
+      cp -drT ${upstream-info.toolchain} native_client/toolchain
+      chmod -R u+w native_client/toolchain
+      find native_client/toolchain -type f -a -perm +0100 -exec "${./patch_toolchain.sh}" {} \;
```
Can't we build the toolchain from source here instead of `patchelf`ing some binaries?

sivertb replied:
We could try that. This would also remove the need for `stdenv_32bit` as you mentioned below. https://www.chromium.org/nativeclient/pnacl/building-pnacl-components-for-distribution-packagers contains some notes on how to build the toolchains. I tried this approach back in June when I first started on this, but couldn't get it to work. It does seem like they've made some improvements since then though, so it might be worth a shot. Sadly there's no separate tar-balls for the toolchains though, we need to run a script supplied with the chromium source

aszlig commented on pkgs/applications/networking/browsers/chromium/common.nix:
```diff
@@ -34,7 +33,7 @@
 buildFun:
 
-with stdenv.lib;
+with stdenv_32bit.lib;
```
Again, I'd like to avoid requiring multilib just for NaCl, if it only works with a 32bit toolchain then we really should separate it.

aszlig commented:
What I really meant with "separate derivation" was not to separate the NaCl toolchain sources but the whole NaCl helper as a separate derivation.

Moreover, shouldn't NaCl be able to be used completely separate from Chromium and also its releases? I mean it has its own bytecode so it should be more or less something like "throw some bytecode at the NaCl black box and wait for something to happen", which shouldn't depend on the browser version. Am I right in this regard?

sivertb commented:
As far as I can tell it's not possible to split out NaCl as a separate component. We can't just combine the NaCl runtime with a chromium binary built without NaCl and expect it to work. If you have a look at e.g. `chrome/chrome_browser.gypi` in the chromium source you'll see that it includes some extra code if NaCl is enabled. The same goes for several of the other gypi files for components that end up in the chromium binary.

sivertb committed:
chromium: Support building with NaCl

This commits adds a new 'config.chromium.enableNaCl' option that makes
it possible to build chromium with native client support enabled.

The NaCl and PNaCl toolchains are required to build chromium with NaCl
support. These are not part of the downloaded tarball, and must be
built as a separate derivation.

'common.nix' linkes the toolchains into place if 'enableNaCl' is true.
It's also been changed to do stripping manually, to avoid stripping the
nacl_irt_*.nexe file, as this results in a segfault when trying to run
NaCl apps.

sivertb committed:
chromium: Update channels

stable: 51.0.2704.103 -> 52.0.2743.82
beta:   52.0.2743.41  -> 53.0.2785.34
dev:    53.0.2767.4   -> 54.0.2810.2

See:
http://googlechromereleases.blogspot.no/2016/07/stable-channel-update.html
http://googlechromereleases.blogspot.no/2016/07/beta-channel-update-for-desktop.html
http://googlechromereleases.blogspot.no/2016/07/dev-channel-update-for-desktop.html

sivertb commented:
I finally got it to work with PNaCl toolchains and translator built from source; no more binary blobs. The NaCl toolchains are also not tied to the chromium version, and the same toolchains can be used to build stable, beta and dev channels.

Most of the information about how to build the PNaCl can be found here:
https://www.chromium.org/nativeclient/pnacl/building-pnacl-components-for-distribution-packagers
https://bugs.chromium.org/p/nativeclient/issues/detail?id=3954

yegortimoshenko commented:
Both NaCl and PNaCl have been deprecated in favor of WebAssembly, except for NaCl on Chromium OS, see:

https://blog.chromium.org/2017/05/goodbye-pnacl-hello-webassembly.html
https://developer.chrome.com/native-client/nacl-and-pnacl

Support will be removed in Q1 2018.
